### ISSUE TITLE
Add a few more combinators

### ISSUE THREAD

**Issue by impinball** [zenparsing/es-observable#58]
I can see there's these methods adapted from the Array prototype:

- `filter -> Observable`
- `forEach -> Observable`
- `map -> Observable`

Couldn't we fill out some of the rest of the methods? It would be really nice and useful, especially for finite observables.

- `concat -> Observable`
    - this observable followed by that observable
- `entries -> Observable<[T, T]>` (like Set's `Iterator<[T, T]>`)
- `every -> Promise<boolean>`
    - if every entry works with that predicate
- `find -> Promise<T>`
- `includes -> Promise<boolean>`
- `join -> Promise<string>`
- `keys -> Observable`
- `reduce -> Promise<T>`
- `some -> Promise<T>`
- `values -> Observable`

These are ones I wouldn't really miss too much because there isn't a lot of applicability for them:

Indices don't make sense with observables.

- `Array.prototype.findIndex()`
- `Array.prototype.indexOf()`
- `Array.prototype.slice()`

These would require at worst case a full evaluation of the observables to be in memory to create a proper counterpart.

- `Array.prototype.lastIndexOf()`
- `Array.prototype.reduceRight()`
- `Array.prototype.reverse()`
- `Array.prototype.sort()`

Mutable Array methods don't make sense with observables.

- `Array.prototype.copyWithin()`
- `Array.prototype.fill()`
- `Array.prototype.pop()`
- `Array.prototype.push()`
- `Array.prototype.shift()`
- `Array.prototype.splice()`
- `Array.prototype.toLocaleString()`
- `Array.prototype.toSource()`
- `Array.prototype.unshift()`

I'm not sure if this is really necessary. If it is, it probably should be comparable to `Array.prototype.join()`.

- `Array.prototype.toString()`

**Comment by ljharb** [comment]
(some small nits: `Array#reverse` and `Array#sort` are mutator methods, and `Array#toSource` isn't in the standard)

**Comment by impinball** [comment]
And that's what I get for copying the list off the side bar of MDN
blindly... :-(

I think you all know what I meant, though. ;-)

**Comment by zenparsing** [comment]
Thanks @impinball , good question.

Originally this proposal only had `forEach` (which returns a promise, BTW), but I found it pretty tough to write compelling event-type things without at least `filter` and `map` being built-in.

I think we'll need to come back to the other array methods a little further down the road.

Some random thoughts looking at the list:

- `concat`:  There are a few different merging strategies which can be applied to observables, owning to the fact that each event has a time associated with it.  I'm not sure if it makes sense to have `concat` without some of the others, like `switchLatest` or `mergeAll`.
- `entries`, `keys`, `values`: I'm not sure how useful these would be in the context of Observable.
- `every`, `some`, `includes`, `reduce`, `join`:  There might be some disagreement over whether they should return a Promise or an Observable of one element.

`slice` might be rather nice.  It would return an Observable of just the elements whose position is within the slice range.

**Comment by impinball** [comment]
I was mostly throwing out ideas. Maybe also a `take` or `drop`?

**Comment by blesh** [comment]
`zip` is probably more important than most people realize.

**Comment by impinball** [comment]
@blesh That is true. I'm almost surprised no one has suggested on es-discuss adding that (and a few others) to the Array prototype.

A polyfill for `zip` specifically could be done using a stack-based buffer + another Observable (for consistency).

**Comment by blesh** [comment]
Well the thing is this, implementing a `zip` or `mergeAll` or `flatMap` is hard, because you end up with unbounded buffers. So then you need to implement buffering strategies, which by itself is nuanced. Do you drop passed a buffer limit? Do you blow up? Do you just let it go until it eats your memory? Do you need to switch between them all?

... that really brings us into a whole realm that I think browser vendors don't want to get into, and I seriously doubt it's something that would be palatable to the TC39 (but perhaps I'm wrong).

I think @zenparsing is doing the right thing by keeping the operators limited.

Although something basic like a `lift` operator might make me happy. ;)

**Comment by impinball** [comment]
@blesh Good point, and I would definitely agree. And even though `flatMap` doesn't require any unbounded memory [1], I would still support keeping the list of primitives at least a little lean. It doesn't negate the point of things like `reduce`, `every`, or `some`.

[1] Here's an example of a `flatMap` for generators. An example for Observables would be algorithmically similar, even for an `Observable<Observable>`.

```js
function *flatMap(iters, f) {
  for (let iter of iters) {
    for (let entry of iter) {
      yield f(entry);
    }
  }
}
```

**Comment by blesh** [comment]
> And even though flatMap doesn't require any unbounded memory.

The thing about flatMap is that you end up with N subscriptions to inner observables. That's where the unbounded memory is... consider the following:

```js
let subscription = Observable.interval(1000)
    .flatMap(x => Observable.interval(10).map(y => x + y))
    .subscribe(::console.log)
```

The above observable will *endlessly* create inner observables that will *never* end. Every single one of those creates an inner subscription via calls to `subscribe` on those inner observables, which are then put on a composite subscription that is returned by the outer `.subscribe`.

**Comment by benjamingr** [comment]
+1 for keeping the proposal minimal and adding operators only after. More operators mean more bikeshedding over the proposal in the TC, we need to get something out so APIs in the DOM can start using it - sugar can (and should) come later.

**Comment by impinball** [comment]
I agree in that we shouldn't rush it. It can be put on hold for this train,
so that the rest is workable by ES7. The extras would definitely sound
great for the ES8 train.

**Comment by domenic** [comment]
I think it's going to be a tough sell to even get filter/map in there given that they're not on Iterator.prototype. I'd be more in support of adding filter/map to both prototypes at once than to only Observable.prototype.

**Comment by benjamingr** [comment]
@domenic are there any plans to add the utility methods to `Iterator.prototype` at this point?

**Comment by domenic** [comment]
@benjamingr work was started on that by @leebyron but IIRC it was pretty firmly derailed during the TC39 meeting he proposed it, so I am not sure what the status is now.

**Comment by zenparsing** [comment]
IIRC, there wasn't much interest in adding methods to `Iterator.prototype`, and the idea of never adding methods floated around (instead relying on adapter-type classes).

We could just leave the proposal at `forEach`, but without `filter` and `map` (a couple of the more commonly used `Array.prototype` methods) it feels pretty anemic.

**Comment by domenic** [comment]
I don't see why Observable.prototype feels any more anemic than Iterator.prototype does. If adapter classes are the answer for the latter, then they should be the answer for the former too.

**Comment by zenparsing** [comment]
I guess iterators have `for-of`, which makes their "zero-tier" (no external requirements) pretty strong.  Adding `filter` and `map` is a way of making the zero-tier for observables a little stronger.

You might be right, though.

**Comment by domenic** [comment]
I guess I was more thinking Map and Set.

**Comment by impinball** [comment]
I also looked at the implementation itself, and I don't think it would be that hard to create a wrapper for it with those methods. I'm looking at the fact [`map`](https://github.com/zenparsing/es-observable/blob/master/src/Observable.js#L361-L381), [`filter`](https://github.com/zenparsing/es-observable/blob/master/src/Observable.js#L383-L403), and [`forEach`](https://github.com/zenparsing/es-observable/blob/master/src/Observable.js#L256-L275) all are completely decoupled from the rest of the polyfill. All three rely exclusively on the public API. I think it may be best to take those off the prototype, and just have an Underscore- or Lodash-like utility library that fufills this use case.

**Comment by RangerMauve** [comment]
@impinball That seems like a good idea, except I think forEach should at least remain to make it similar to other "collections" like Map and Set.

**Comment by leebyron** [comment]
IIRC the concern with the last proposed additions to Iterator.prototype resulted from an unsatisfactory approach in creating reasonable behavior that would support all of Map/Set iterables, feel at home with Array prototype, Iterator instances, and non-built-in Iterator and Iterables. There was some interest in exploring an "adaptor pattern", and I'm still interested in figuring it out, so it's not a dead-end yet.

**Comment by impinball** [comment]
@RangerMauve Good point, although you could argue against in that `@@iterator` isn't defined for it. That could go either way.

**Comment by zenparsing** [comment]
For many of the same reasons that we want platform libraries to use promises over callbacks, we also (presumably) want platform libraries to use observables over callbacks (for push-based event thingies anyway).

From my perspective, if Observable doesn't have *anything* except subscribe, that's a very hard sell.  If it has only `forEach`, it's a better sell but not great.  Throw `filter` and `map` in there and I think it becomes compelling.

That said, I'm fine with dropping `filter` and `map` if that helps the process.  `forEach` must stay, though.

**Comment by impinball** [comment]
A little gist featuring forEach, map, filter, and reduce: https://gist.github.com/impinball/4fbb6a0f093d0a97571c

**Comment by zenparsing** [comment]
@impinball cool.

Note that Symbol.observable should be a method and not a getter (as with Symbol.iterator).  And the core proposal with definitely have `forEach` : )

Another way to do this kind of extension would be via subclassing.  Then instead of a wrapping function you should be able to just use `from`:

```js
class MyObservable extends Observable {
    reduce(f, start) {
        // implementation...
    }
    // All my other cool combinator methods
}

MyObservable.from(someOtherObservable).reduce(...);
```

If you're interested, it would be really cool to have some kind of lightweight combinator library over this core proposal.  I think it would strengthen this proposal.

**Comment by blesh** [comment]
> I don't see why Observable.prototype feels any more anemic than Iterator.prototype does. If adapter classes are the answer for the latter, then they should be the answer for the former too.

I agree with @domenic here. But I think I'd like to see `forEach`, `filter` and `map`, etc on Iterators personally. It's weird to me that they don't exist, "Here's this powerful collection primitive, but if you want to do anything cool with it, you better role your own basic methods"

... but that's a totally different issue.

**Comment by impinball** [comment]
@blesh 

Node.js and WHATWG Streams suffered the same treatment. There are wrapper libraries out there full of those basic methods for those and Generators. Arrays were the same way, and they had Underscore early on before `map`/`filter`/etc. made it into the standard. Promises/A+ promises were almost ubiquitous when ES6 promises came out, and the ES6 spec itself is almost a Promises/A+ conforming implementation.

Generally, the best way to introduce new features like this is to keep the initial opinions out of the standard, and when the feature matures, add what everyone has already started using.

And hey, at least we're not still writing ES3! :wink:

**Comment by blesh** [comment]
To clarify, I think for now, if it were up to me (which it's not), I'd opt to exclude any operators and just let libraries implement those.

**Comment by RangerMauve** [comment]
To be honest, I agree with @blesh on this. However I think that there should be a lot of examples of how one could subclass Observable in order to add support for any of these fancy chaining methods. If there isn't a clear path for how libraries should wrap over Observable and work with it, then I think that having more methods would be more important, otherwise user adoption will be bad since people will be expecting the fancy stuff around RxJS or BaconJS, but they'll be getting something super low level and more tedious to use. If anything, I'd expect users to just stick to libraries and ignore the native Observable altogether.

**Comment by impinball** [comment]
@RangerMauve 

Would something like this do the job (note: untested and not optimized)? It's actually pretty straightforward, since Observables are easy to subclass. I can make an npm module out of it later if you all would like.

```js
// I used decorators to simplify the implementation a lot.
function subscribe(that, f) {
    return new that.constructor[Symbol.species](observer => {
        const unsubscribe = that.subscribe(f(observer, () => unsubscribe()))
    })
}

function run(that, f) {
    return subscribe(that, (observer, unsubscribe) => ({
        next(value) { return f(value, observer, unsubscribe) },
        error(value) { return observer.error(value) },
        complete(value) { return observer.complete(value) },
    })
}

function descriptor(f) {
    return (target, name, desc) => {
        const g = f(desc.value)
        desc.value = function (...args) { return g(this, ...args) }
    }
}

const check = descriptor(old => (that, f, ..rest) => {
    if (typeof f !== 'function') {
        throw new TypeError('Expected f to be a function')
    }
    return old.call(that, f, ...rest)
})

const wrap = descriptor(old => (that, f) => run(that, old.bind(that, f)))

export default class Wrapper extends Observable {
    @check forEach(f) {
        return new Promise((resolve, reject) => subscribe(this, observer => ({
            next(value) {
                try { f(value) }
                catch (e) { return observer.error(e) }
                return observer.next(value)
            },
            error: reject,
            complete: resolve,
        })))
    }

    @check @wrap tap(f, value, observer) {
        try { f(value) }
        catch (e) { return observer.error(e) }
        return observer.next(value)
    }
  
    @check @wrap map(f, value, observer) {
        try { value = f(value) }
        catch (e) { return observer.error(e) }
        return observer.next(value)
    }
  
    @check @wrap filter(f, value, observer) {
        try { if (!f(value)) return undefined }
        catch (e) { return observer.error(e) }
        return observer.next(value)
    }

    reject(f) {
        return this.filter(x => !f(x))
    }

    take(n) {
        n = +n
        return run(this, (value, observer, unsubscribe) => {
            if (n <= 0) return unsubscribe()
            n--
            return observer.next(value)
        })
    }

    drop(n) {
        n = +n
        return run(this, (value, observer) => ({
            if (n <= 0) return observer.next(value)
            n--
        }))
    }

    @check @wrap takeWhile(f, value, observer, unsubscribe) {
        try { if (!f(value)) return unsubscribe() }
        catch (e) { return observer.error(e) }
        return observer.next(value)
    }

    @check dropWhile(f) {
        let done = false
        return run(this, (value, observer) => ({
            if (!done) {
                try { if (done = f(value)) return undefined }
                catch (e) { return observer.error(e) }
            }
            return observer.next(value)
        }))
    }
    
    @check reduce(f, start) {
        return new Promise((resolve, reject) => {
            let res = start
            return subscribe(this, observer => ({
                next(value) {
                    try { res = f(res, value) }
                    catch (e) { return observer.error(e) }
                    return observer.next(value)
                },
                error: reject,
                complete() { return resolve(res) },
            }))
        })
    }
}
```

**Comment by impinball** [comment]
Also, +infinity for the whole proposal itself. I myself need a lightweight, low-level stream/observable that works with objects and propogates errors. Node streams don't propogate errors, and WHATWG streams aren't lightweight and low-level enough.

This proposal is exactly what I need, so I'm increasingly tempted to copy large chunks of the polyfill so I can use the parts I need now.

**Comment by RangerMauve** [comment]
@impinball Yeah, something like that would be pretty cool. :o

**Comment by leebyron** [comment]
Coming back to this now as I'm starting to revive a proposal for IteratorPrototype additions, and I agree with the desire to have a small set of base higher-order prototype methods. Awesome to see filter/map already in the polyfill code, `reduce(fn): Promise` would be a desirable addition to round out the basic set.

**Comment by zenparsing** [comment]
@leebyron Sounds good.  I think `reduce` would be a very useful addition as well.

**Comment by benjamingr** [comment]
Reduce as a better forEach, lots of the RxJS combinators and higher order methods are super useful and I think every Rx user uses at least 10 of them all the time. 

A lot of the power of observables comes from things like `flatMapLatest`, I think not including these methods definitely cripples observables, I'd just like to see us get _something_ in first people can base APIs on and then extend it.

**Comment by RangerMauve** [comment]
I think it might be important to just get the low level implementation details ironed out first so that existing observable libraries can be inter-operable quicker. I personally see the observable spec not as a replacement of RxJS and the such, but as a way to get them to talk to work well with each other. Sort of like how Promises/A+ mostly helped get all the promise libraries on the same page so that they could be used together. To that end, I almost feel like having `forEach` isn't even necessary.

**Comment by RangerMauve** [comment]
Specifically, I think it'd make sense if all observable libraries implemented `Observable.from()` to convert anything with `Symbol.observable` into their own flavor of Observable. And just made sure that they used it whenever they had a function which took an observable as an argument. Then if they all implement `Symbol.observable` in a standard way, they would all play nicely together.

**Comment by benjamingr** [comment]
Can we close this?

**Comment by zenparsing** [comment]
I think we can.

**Comment by RangerMauve** [comment]
Just to confirm, what is the full list of combinators that's being included in the spec so far?

**Comment by blesh** [comment]
@RangerMauve ... my understanding is _zero_ combinators are included.

**Comment by RangerMauve** [comment]
@blesh Perfect. :D
